src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/Unsafe.cs

```diff
@@ -667,7 +667,7 @@ public static void Write<T>(void* destination, T value)
         // Mono:AsRef
         [NonVersionable]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static ref T AsRef<T>(scoped in T source)
+        public static ref T AsRef<T>(scoped ref readonly T source)
```

Comment:
For .NET 9, I think it may be worth a discussion in API review on whether we want more of `Unsafe` to take `ref readonly`.

The general consideration is that most APIs on `Unsafe` are "unsafe" and many, such as `Add`, do not actually mutate the contents of the `ref` but rather simply adjust the `ref` to point to a new address.

If we were to changes these APIs to be, for example, `ref T Add(ref readonly T source, int elementOffset)`, then it basically has an implicit `ref T AsRef<T>(ref readonly T source)` built in.

The benefit of this is that it reduces "clutter" and can improve readability of such `Unsafe` code.  Consider the simple case below:
```diff
- ref readonly T address = ref Unsafe.Add(ref Unsafe.AsRef(in source), (nint)elementOffset);
+ ref readonly T address = ref Unsafe.Add(in source, (nint)elementOffset);
```

The downside is that any `Unsafe` operation can go from `mutable` to `immutable` ref and users could introduce a mutation by accident. This could be handled with an analyzer though, suggesting users preserve mutability/immutability when using `Unsafe` and to insert `Unsafe.AsRef` explicitly if it was intentional.

In an ideal world, we might have overloads that were `ref T Add(ref T source, int offset)` and `ref readonly T Add(ref readonly T source, int offset)`. However, we can't due to them both being `T&` in IL. Alternatively we could request a language feature that basically says "I match the mutability of my input". Such a feature would also be beneficial for some ref returns. But an analyzer seems like an acceptable solution to this problem instead